File: src/c/magma/split/StatementSplitter.c
import magma.collect.List;
import magma.java.JavaLinkedList;
import magma.java.JavaList;
import magma.java.Strings;
struct StatementSplitter implements Splitter {
	StringBuilder merge(StringBuilder inner, String compiled){
		return inner.append(compiled);
	}
	List<String> split(String input){
		auto segments = JavaList<String>();
		auto buffer = StringBuilder();
		auto depth = 0;
		auto queue = Strings.streamChars(input).collect(JavaLinkedList.collector());
		while (!queue.isEmpty()) {
			auto c = queue.popOrPanic();
			buffer.append(c);
			if (c == '\'') {
				auto popped = queue.popOrPanic();
				buffer.append(popped);
				if (popped == '\\') {
					buffer.append(queue.popOrPanic());
				}
				buffer.append(queue.popOrPanic());
				continue;
			}
			if (c == '"') {
				while (!queue.isEmpty()) {
					auto next = queue.popOrPanic();
					buffer.append(next);
					if (next == '"') {
						break;
					}
					if (next == '\\') {
						buffer.append(queue.popOrPanic());
					}
				}
			}
			if (c == ';' && depth == 0) {
				Splitter.advance(segments, buffer);
				buffer = new StringBuilder();
			}
			else {
				if (c == '}' && depth == 1) {
					depth--;
					Splitter.advance(segments, buffer);
					buffer = new StringBuilder();
				}
			}
			else {
				if (c == '{' || c == '(') {
					depth++;
				}
				if (c == '}' || c == ')') {
					depth--;
				}
			}
		}
		Splitter.advance(segments, buffer);
		return segments;
	}
}

File: src/java/magma/split/StatementSplitter.java
package magma.split;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

public class StatementSplitter implements Splitter {
    @Override
    public StringBuilder merge(StringBuilder current, String element) {
        return current.append(element);
    }

    @Override
    public List<String> split(String content) {
        final var segments = new ArrayList<String>();
        var buffer = new StringBuilder();
        var depth = 0;

        final var queue = IntStream.range(0, content.length())
                .mapToObj(content::charAt)
                .collect(Collectors.toCollection(LinkedList::new));

        while (!queue.isEmpty()) {
            final var c = queue.pop();
            buffer.append(c);

            if (c == '\'') {
                final var c1 = queue.pop();
                buffer.append(c1);

                if (c1 == '\\') buffer.append(queue.pop());
                buffer.append(queue.pop());
            }

            if (c == '"') {
                while (!queue.isEmpty()) {
                    final var c1 = queue.pop();
                    buffer.append(c1);

                    if (c1 == '"') break;
                    if (c1 == '\\') buffer.append(queue.pop());
                }
            }

            if (c == ';' && depth == 0) {
                advance(segments, buffer);
                buffer = new StringBuilder();
            } else if (c == '}' && depth == 1) {
                depth--;
                advance(segments, buffer);
                buffer = new StringBuilder();
            } else {
                if (c == '{') depth++;
                if (c == '}') depth--;
            }
        }

        advance(segments, buffer);
        if (depth != 0) {
            System.err.println("Invalid depth: '" + depth + "': " + content);
        }

        return segments;
    }
}
